tikv-client/src/main/java/com/pingcap/tikv/meta/collate/GeneralCICollator.java

```diff
@@ -0,0 +1,421 @@
+/*
+ * Copyright 2022 PingCAP, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.pingcap.tikv.meta.collate;
+
+import com.pingcap.tikv.meta.Collation;
+import com.pingcap.tikv.util.Pair;
+import java.io.ByteArrayOutputStream;
+
+public class GeneralCICollator {
+
+  // if a > b return 1, if a < b return -1, if a == b return 0
+  public static int compare(String a, String b) {
+    a = Collation.truncateTailingSpace(a);
+    b = Collation.truncateTailingSpace(b);
+    int r1 = 0, r2 = 0;
+    int ai = 0, bi = 0;
+    while (ai < a.length() && bi < b.length()) {
+      Pair<Integer, Integer> pair1 = Collation.decodeRune(a, ai);
+      r1 = pair1.first;
+      ai = pair1.second;
+      Pair<Integer, Integer> pair2 = Collation.decodeRune(b, bi);
+      r2 = pair2.first;
+      bi = pair2.second;
+      int cmp = convertRuneGeneralCI(r1) - convertRuneGeneralCI(r2);
+      if (cmp != 0) {
+        return Collation.sign(cmp);
+      }
+    }
+    return Collation.sign((a.length() - ai) - (b.length() - bi));
+  }
+
+  // truncate tail space and encode string to bytes
+  public static byte[] key(String key) {
+    return keyWithoutTrimRightSpace(Collation.truncateTailingSpace(key));
+  }
+
+  public static byte[] keyWithoutTrimRightSpace(String key) {
+    ByteArrayOutputStream buf = new ByteArrayOutputStream();
+    int i = 0, r = 0;
+    while (i < key.getBytes().length) {
+      Pair<Integer, Integer> pair = Collation.decodeRune(key, i);
+      r = pair.first;
+      i = pair.second;
+      int u32 = convertRuneGeneralCI(r);
+      buf.write((byte) ((u32 & 0xFFFF) >> 8));
+      buf.write((byte) (u32 & 0xFF));
+    }
+    return buf.toByteArray();
+  }
+
+  private static int convertRuneGeneralCI(int r) {
+    if (r > 0xFFFF) {
+      return 0xFFFD;
+    }
+    int[] plane = planeTable[r >> 8];
+    if (plane == null) {
+      return r;
+    }
+    return plane[r & 0xFF];
+  }
+
+  private static final int[] plane00 =
```

Comment:
can you add some comment here to explain what these plans mean?

Reply:
Sorry, I don't understand encoding, too. And I referred from [TiDB general_ci](https://github.com/pingcap/tidb/blob/a33d971adde7e5750cbddb2db2664b3389b0ab69/util/collate/general_ci.go). Maybe you can find some solutions here.